### PR TITLE
Add filter to track 5xx errors for Prometheus

### PR DESCRIPTION
**Pull request by mkuznyetsov**
### What does this PR do?
Add metric counter "tomcat.server.errors" for counting Tomcat server error responses (with 5xx HTTP status), that would be exposed for Prometheus

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/12244

### CODE REVIEW

**Comment by metlos** [comment]
I think the metric should be called something along the lines of `che.server.errors`, because it doesn't measure 500 errors in tomcat as a whole but specifically just in wsmaster.

**Comment by skabashnyuk** [comment]
what if 
```
che.server.api.response{"area"="http", "code=500"} 500
che.server.api.response{"area"="http", "code=400"} 233
che.server.api.response{"area"="http", "code=200"} 445
che.server.api.response{"area"="ws", "code=400"} 435223
```
Like this one

**Comment by mkuznyetsov** [comment]
@metlos yes, I can go with that name.
@skabashnyuk do you want to have different counters for 5xx, 4xx, 3xx, 2xx, basically counters for all statuses? Or even have each counter for specific http statuses, like 500, 501, 502, 503 etc.

can we agree on finite amount of counters (and tags related to them) that we want to track?

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+      throws IOException, ServletException {
+    filterChain.doFilter(servletRequest, servletResponse);
+    if (servletResponse instanceof HttpServletResponse) {
+      HttpServletResponse hts = (HttpServletResponse) servletResponse;
+      if (hts.getStatus() / 100 == 2) {
+        apiResponseCounter.incrementSuccessResponseCounter();
+      }
+      if (hts.getStatus() / 100 == 3) {
+        apiResponseCounter.incrementRedirectResonseCounter();
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: typo in word `Resonse`

Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: thanks

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+  public void init(FilterConfig filterConfig) throws ServletException {}
+
+  @Override
+  public void doFilter(
+      ServletRequest servletRequest, ServletResponse servletResponse, FilterChain filterChain)
+      throws IOException, ServletException {
+    filterChain.doFilter(servletRequest, servletResponse);
+    if (servletResponse instanceof HttpServletResponse) {
+      HttpServletResponse hts = (HttpServletResponse) servletResponse;
+      if (hts.getStatus() / 100 == 2) {
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: You get status and divide 4 times. Consider saving the result in a variable and maybe using switch statement.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  private Counter redirectResponseCounter;
+  private Counter clientErrorResponseCounter;
+  private Counter serverErrorResponseCounter;
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+    successResponseCounter =
+        Counter.builder("che.server.api.response.success")
+            .description("Che Server Tomcat success responses (2xx responses)")
+            .tags("code=200", "area=http")
```
Comment (skabashnyuk) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: 2xx

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+  @Override
+  public void doFilter(ServletRequest request, ServletResponse response, FilterChain filterChain)
+      throws IOException, ServletException {
+    filterChain.doFilter(request, response);
+    if (response instanceof HttpServletResponse) {
+      HttpServletResponse httpResponse = (HttpServletResponse) response;
+      int status = httpResponse.getStatus() / 100;
+      switch (status) {
+        case 2:
```
Comment (skabashnyuk) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: move switch to apiResponseCounter

File: wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java [hunk 5]
```diff
@@ -223,4 +223,10 @@ public void remove(@ApiParam("User identifier") @PathParam("id") String id)
   private static String userId() {
     return EnvironmentContext.getCurrent().getSubject().getUserId();
   }
+
```
Comment (skabashnyuk) [inline wsmaster/che-core-api-user/src/main/java/org/eclipse/che/api/user/server/UserService.java]: don't forget to remove this change

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+ *   Red Hat, Inc. - initial API and implementation
+ */
+package org.eclipse.che.core.metrics;
+
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.binder.MeterBinder;
+import javax.inject.Singleton;
+
+@Singleton
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: I think it is better to have an annotation just above class definition but not java doc

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+/**
+ * Metrics
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: So terse java docs )

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponceMetricFilterTest.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+  @InjectMocks private ApiResponceMetricFilter filter;
+
+  @BeforeMethod
+  public void setUp() {
+    filter = new ApiResponceMetricFilter();
+  }
+
+  @Test
+  public void shouldIncrementCounterOnSuccessResponses(int status) {
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponceMetricFilterTest.java]: I don't see data provider for this test. Does it work?

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.search.RequiredSearch;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Test;
+
+public class ApiResponseCounterTest {
+  ApiResponseCounter apiResponseCounter;
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java]: Why do you need package private fields? May they be private?

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+  @Test
+  public void testServerError() {
+    apiResponseCounter.handleStatus(500);
+
+    RequiredSearch meter = registry.get("che.server.api.response.success");
+    registry.getMeters();
+  }
+
+  @DataProvider(name = "information")
+  public Object[][] information() {
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java:95]: You are going to use these data providers somewhere in tests?

File: multiuser/permission/che-multiuser-permission-user/src/main/java/org/eclipse/che/multiuser/permission/user/UserServicePermissionsFilter.java [hunk 11]
```diff
@@ -54,6 +54,7 @@ protected void filter(GenericResourceMethod genericResourceMethod, Object[] argu
       case "updatePassword":
       case "getById":
       case "find":
+      case "test":
```
Comment (sleshchenko) [inline multiuser/permission/che-multiuser-permission-user/src/main/java/org/eclipse/che/multiuser/permission/user/UserServicePermissionsFilter.java]: Please revert this change

**Review by sleshchenko — COMMENTED** [review]
Just a few more steps to get PR ready to merge

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponceMetricFilterTest.java [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+import org.everrest.assured.EverrestJetty;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.testng.MockitoTestNGListener;
+import org.testng.annotations.BeforeMethod;
+import org.testng.annotations.Listeners;
+import org.testng.annotations.Test;
+
+@Listeners({EverrestJetty.class, MockitoTestNGListener.class})
+public class ApiResponceMetricFilterTest {
```
Comment (metlos) [inline core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponceMetricFilterTest.java]: Please fix the typo in the name. Should be `ApiResponseMetricFilterTest`.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.FilterConfig;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.http.HttpServletResponse;
+
+@Singleton
+public class ApiResponceMetricFilter implements Filter {
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: Please fix the typo in the name. Should be `ApiResponseMetricFilter`.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+
+  @Inject protected ApiResponseCounter apiResponseCounter;
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: Other places in the codebase seem to prefer constructor injection with private final fields. Is it also possible for the servlet filters?

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+public class ApiResponseCounter implements MeterBinder {
+  protected Counter informationalResponseCounter;
+  protected Counter successResponseCounter;
+  protected Counter redirectResponseCounter;
+  protected Counter clientErrorResponseCounter;
+  protected Counter serverErrorResponseCounter;
+
+  @Override
+  public void bindTo(MeterRegistry registry) {
+    informationalResponseCounter =
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java:39]: I'd personally prefer having a single metric - `che.server.api.response` with different tag values.
E.g. you never have a `che.server.api.response.success` with any other code than `2xx`. That makes the tag useless (unless we also use that tag in other areas, which I doubt makes much sense).

So instead of having 5 different metrics, each with a unique code tag, I'd make a single metric with 5 different code tag values. This also makes it possible to easily aggregate the responses with e.g. 2xx and 3xx status codes.

Comment (metlos) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java:39]: This can be done by just changing the counters to have the same name.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        Counter.builder("che.server.api.response.informational")
+            .description("Che Server Tomcat informational responses (1xx responses)")
+            .tags("code=1xx", "area=http")
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: Should be `.tags("code", "1xx", "area", "http")` or `.tag("code", "1xx").tag("area", "http")`

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+            .register(registry);
+    successResponseCounter =
+        Counter.builder("che.server.api.response.success")
+            .description("Che Server Tomcat success responses (2xx responses)")
+            .tags("code=2xx", "area=http")
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: See my comment above for the correct way of declaring the tags.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+            .register(registry);
+    redirectResponseCounter =
+        Counter.builder("che.server.api.response.redirect")
+            .description("Che Server Tomcat redirect responses (3xx responses)")
+            .tags("code=3xx", "area=http")
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: See my comment above for the correct way of declaring the tags.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+            .register(registry);
+    clientErrorResponseCounter =
+        Counter.builder("che.server.api.response.client.error")
+            .description("Che Server Tomcat client errors (4xx responses)")
+            .tags("code=4xx", "area=http")
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: See my comment above for the correct way of declaring the tags.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+            .register(registry);
+    serverErrorResponseCounter =
+        Counter.builder("che.server.api.response.server.error")
+            .description("Che Server Tomcat server errors (5xx responses)")
+            .tags("code=5xx", "area=http")
```
Comment (metlos) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: See my comment above for the correct way of declaring the tags.

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 16 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: ok

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 17 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: ok

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 18 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: ok

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 20 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: ok

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 19 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java]: ok

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 15 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java:39]: alright

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java (hunk 14 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: I saw in other filters there is also method or field injection. And I had to do it when it come to running rest-assured tests that couldn't instantiate a filter that would have constructor with arguments

Comment (metlos) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponceMetricFilter.java]: :+1: thanks for the explanation. Maybe a comment at the field explaining this would help the casual observer in the future ;)

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java (hunk 10 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseCounterTest.java:95]: yes, tests are now finished

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        Counter.builder("che.server.api.response")
+            .description("Che Server Tomcat server errors (5xx responses)")
+            .tag("code", "5xx")
+            .tag("area", "http")
+            .register(registry);
+  }
+
+  public void handleStatus(int status) {
+    status = status / 100;
+    switch (status) {
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java:73]: Does maven build output contain warning that there is switch without `default` section?

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+
+  @InjectMocks private ApiResponseMetricFilter filter;
+
+  @BeforeMethod
+  public void setUp() {
+    filter = new ApiResponseMetricFilter();
+  }
+
+  @Test
+  public void shouldIncrementCounterOnSuccessResponses(int status) {
```
Comment (sleshchenko) [inline core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java]: I don't see data provider for this test. What the value is used for `status` here?

**Review by sleshchenko — APPROVED** [review]
LGTM

Please take a look my inline comments

**Review by metlos — APPROVED** [review]
LGTM

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java (hunk 22 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java]: fixed the name of the test, it just tests the invokation of ApiResponseCounter's method, and it's logic is tested in a different test

File: core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java (hunk 21 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/main/java/org/eclipse/che/core/metrics/ApiResponseCounter.java:73]: there are no warnings, but I can add default section with comment for clarity

**Comment by mkuznyetsov** [comment]
ci-test

**Review by skabashnyuk — APPROVED** [review]

File: core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java (hunk 22 above)
Comment (mkuznyetsov) [reply core/che-core-metrics-core/src/test/java/org/eclipse/che/core/metrics/ApiResponseMetricFilterTest.java]: fixed the wrongfully written test, that was actually skipped during the build

**Comment by mkuznyetsov** [comment]
ci-build

**Comment by mkuznyetsov** [comment]
ci-build
